infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java

```diff
@@ -346,7 +347,18 @@ static void restartServer(boolean resetData) {
     private static void cleanShutdown() {
         log.info("CleanShutdown: Starting Cleanup.");
         shutdownServer = true;
-        activeServer.close();
+        try {
+            CorfuServerNode current = activeServer;
+            if (current != null) {
+                activeServer.close();
+            }
+        } catch (Throwable th) {
+            log.error("cleanShutdown: failed during shutdown", th);
+        }
+
+        // Flush the async appender before exiting to prevent the loss of logs
+        LoggerContext loggerContext = (LoggerContext) LoggerFactory.getILoggerFactory();
```

Comment:
This doesn't help at all, I suspect it can even cause loss of the logs.
Logback already have solved the problem, and we have it enabled https://github.com/CorfuDB/CorfuDB/blob/30c8d154f59dfe22d4cc16b94ab9e2c51851c2e5/infrastructure/src/main/resources/logback.prod.xml#L5

Reply:
This is not the Matrix, but things are not what they seem to be.

The shutdown hook configuration added in #2049 is malformed and actually doesn't register a hook that stops the logger's context on shutdown. This is true despite what logback's documentation says, it might be outdated, there might be a bug, or it might just be inaccurate. At least for version 1.2.3 it doesn't work.

For a shutdown hook to be called, a shutdown thread must be registered with `Runtime.getRuntime().addShutdownHook` for it to be executed. If you run the server and collect a heap dump, you can inspect all the shutdown hooks registered and you won't see any hooks registered by logback. This means that the logger context is not being stopped properly on shutdown, which explains some of the log loss I've seen debugging a couple of bugs. 

Now even if that configuration works, its not the best way of implementing "shutdown" logic for a logger because there is no order on shutdown hooks. This means that logging statements produced during the execution of the shutdown hook can be lost this is due to an undefined execution order of the hooks. If the logback shutdown hook is executed first  then subsequent logs will be lost from other hooks.

If you are not convinced, then you can inspect logback's source code here: https://github.com/qos-ch/logback/blob/a154cd1b564d436c90a26b8cb1a2e8ffff0a4a47/logback-core/src/main/java/ch/qos/logback/core/joran/action/ShutdownHookAction.java#L46

If you inspect the code snippet I linked you'll see that it tries to find a `class` attribute, but fails to do so (because it doesn't exist). This means that even the "default" hook isn't being registered. 

Furthermore, the documentation you link about the default hook is incorrect, if you checkout logback's code base version 1.2.3 you'll see that a `DefaultShutdownHook` type doesn't even exist.

Finally, to fix the configuration bug introduced in #2049  the class and full type path must be supplied. For example, the correct configuration for installing a default hook would be

 `<shutdownHook class="ch.qos.logback.core.hook.DelayingShutdownHook"/>`


Anyways, I think this PR fixes an existing problem on master and improves logging in general (via explicit shutdown ordering) and I think we should go forward with it.

Reply:
So,
- let's add DelayingShutdownHook or any other class that is required to the shutdownHook?
- what I see from source code is all the hooks are run in parallel, which means no hooks wait for each other. If you do not shutdown JVM then no reason for shutting down the logger. If the JVM gonna fail you will lose your logs, either way, it doesn't matter if you shutdown it by a hook or explicitly in your code, the effect will be the same.

I still don't see the motivation for adding the code that shouldn't be added. because we have a solution provided by the library of the box.
That's great if you found a misconfiguration bug in the config, let's fix the config?

![image](https://user-images.githubusercontent.com/314292/87489422-d5a25100-c5f7-11ea-854b-4dde21e28159.png)

Reply:
> I still don't see the motivation for adding the code that shouldn't be added


Please read my response again. 

Even if you ignore the logback bug, the PR doesn't re-implement the "out-of-box" behavior. This PR adds execution order which I already describe in detail.


>  hooks are run in parallel, which means no hooks wait for each other.


I never claimed that. I explicitly mention **undefined execution order** , which is the whole point of having one hook and having serial order, as opposed to having two hooks and no order.

Reply:
Ok, I see your point